src/DessserQCheck.ml

```diff
@@ -172,72 +172,50 @@ let shrink_mac_type mt =
   loop to_simplest
 
 let rec shrink_value_type =
+  let open Iter in
   let vt_of_mn mn = mn.T.vtyp in
-  let shrink_fields mns make_typ f =
-    Array.iter (fun (_, mn) -> shrink_maybe_nullable mn (f % vt_of_mn)) mns ;
-    let shrink_mns =
-      let shrink (fn, mn) =
-        Iter.map (fun mn -> fn, mn) (shrink_maybe_nullable mn) in
-      Shrink.filter (fun mns -> Array.length mns > 1)
-        (Shrink.array ~shrink) mns |>
-      Iter.map make_typ in
-    shrink_mns f in
+  let shrink_fields f = (Shrink.pair Shrink.nil shrink_maybe_nullable) f in
+  let non_empty_arrays = (fun mms -> Array.length mms > 1) in
   function
-  | T.Unknown | T.Ext _ ->
-      Iter.empty
-  | T.Base Unit ->
-      Iter.empty
-  | T.Base mt ->
-      (fun f ->
-        shrink_mac_type mt (fun mt -> f (T.Base mt)))
-  | T.Usr _ ->
-      Iter.empty
+  | T.Unknown | T.Ext _ -> empty
+  | T.Base Unit -> empty
+  | T.Base mt -> shrink_mac_type mt >|= T.base
+  | T.Usr _ -> empty
   | T.Vec (dim, mn) ->
-      (fun f ->
-        shrink_maybe_nullable mn (fun mn ->
-          f mn.vtyp ;
-          f (T.Vec (dim, mn))))
+    let smn = shrink_maybe_nullable mn in
+    (smn >|= T.vec dim) <+> (smn >|= vt_of_mn)
   | T.Lst mn ->
-      (fun f ->
-        shrink_maybe_nullable mn (fun mn ->
-          f (T.Lst mn) ;
-          f mn.vtyp))
+    let smn = shrink_maybe_nullable mn in
+    (smn >|= T.lst) <+> (smn >|= vt_of_mn)
   | T.Set (st, mn) ->
-      (fun f ->
-        shrink_maybe_nullable mn (fun mn ->
-          f (T.Set (st, mn)) ;
-          f mn.vtyp))
+    let smn = shrink_maybe_nullable mn in
+    (smn >|= T.set st) <+> (smn >|= vt_of_mn)
   | T.Tup mns ->
-      (fun f ->
-        Array.iter (fun mn -> shrink_maybe_nullable mn (f % vt_of_mn)) mns ;
-        let shrink_mns =
-          Shrink.filter (fun mns -> Array.length mns > 1)
-            (Shrink.array ~shrink:shrink_maybe_nullable) mns |>
-          Iter.map (fun mns -> T.Tup mns) in
-        shrink_mns f)
+      Shrink.filter non_empty_arrays (
+        Shrink.array ~shrink:shrink_maybe_nullable) mns >|= T.tup
   | T.Rec mns ->
-      shrink_fields mns (fun mns -> T.Rec mns)
+      Shrink.filter non_empty_arrays (
+        Shrink.array ~shrink:shrink_fields) mns >|= T.record
   | T.Sum mns ->
-      shrink_fields mns (fun mns -> T.Sum mns)
+      Shrink.filter non_empty_arrays (
+        Shrink.array ~shrink:shrink_fields) mns >|= T.sum
   | T.Map (k, v) ->
-      (fun f ->
-        shrink_maybe_nullable k (f % vt_of_mn) ;
-        shrink_maybe_nullable v (f % vt_of_mn) ;
-        let shrink_kv =
-          (Shrink.pair shrink_maybe_nullable shrink_maybe_nullable) (k, v) |>
-          Iter.map (fun (k, v) -> T.Map (k, v)) in
-        shrink_kv f)
+      let sk = shrink_maybe_nullable k in
+      let sv = shrink_maybe_nullable v in
+      (sk >|= vt_of_mn) <+> (sv >|= vt_of_mn)
+        <+>
+      (pair sk sv >|= fun (k, v) -> T.map k v)
```

Comment:
I think there is the same issue here, although smileys and tie-fighters makes it harder to read for me

Reply:
This one is ok. `sk >|= vt_of_mn` generates for the key and `sv >|= vt_of_mn` for the values.